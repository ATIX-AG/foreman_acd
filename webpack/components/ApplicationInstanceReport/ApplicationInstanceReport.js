import React, { useState } from 'react'
import PropTypes from 'prop-types';

import {
  Icon,
  Spinner,
} from 'patternfly-react';

import {
  VerticalTabs,
} from 'patternfly-react-extensions';

import {
  translate as __
} from 'foremanReact/common/I18n';

import PowerStatus from 'foremanReact/components/hosts/powerStatus';
import ReportViewer from './components/ReportViewer';

class ApplicationInstanceReport extends React.Component {

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {
      data: { hosts, deploymentState, },
      initApplicationInstanceReport,
      setActiveHost,
    } = this.props;

    initApplicationInstanceReport(hosts, deploymentState);
  };

  isActive(id) {
    return (this.props.activeHostId === id);
  }

  collectLastReportData(hosts) {
    const {
      setActiveHost,
    } = this.props;

    const tabs = []
    for (const [index, value] of hosts.entries()) {
      tabs.push(
        <VerticalTabs.Tab
          id={index}
          key={"vt_tab_"+index}
          title={ value.name }
          wrapStyle='nowrap'
          onActivate={() => setActiveHost(index)}
          active={this.isActive(index)}
        />
      );
    }

    return tabs;
  }

  lastReportStatus(host) {
    if (!host['id']) {
      return (
        <div>
          <span>Host: { host['name'] }</span>
          <span>&nbsp;|&nbsp;</span>
          <span>State: { __('not build') }</span>
          <span>&nbsp;|&nbsp;</span>
          <span>Power Status: { __('unknown') }</span>
        </div>
      )
    } else {
      const already_deployed_msg = __("Already existing host which was added to the application instance");
      return (
        <div>
          <span>Host: <a href={ host['hostUrl'] }>{ host['name'] }</a></span>
          <span>&nbsp;|&nbsp;</span>
          <span>State: { host['build'] == true ? "in Build" : "Deployed" }</span>
          <span>&nbsp;|&nbsp;</span>
          <span>Power Status: <PowerStatus key={ "power_status_"+ host['id'] } data={{ id: host['id'], url: host['powerStatusUrl'] }} /></span>
          {host['isExistingHost'] ? (
            <span>
              &nbsp;|&nbsp; Existing host &nbsp;
              <Icon style={{marginRight: 8, marginLeft: 2}} type="pf" name="info" title={already_deployed_msg} />
            </span>
          ) : (<span></span>)}
        </div>
      )
    }
  }

  render() {
    const {
      data: { appInstanceId, appInstanceName, deployTaskUrl, configureJobUrl, reportDataUrl },
      activeHostId, hosts, deploymentState,
      loadReportData,
    } = this.props;

    let tabs = [];
    let reportStatus = undefined;
    let report = undefined;

    // This handles the first call to render() in which the state hosts is always empty
    if (hosts.length == 0) {
      return (<span>No host</span>);
    }

    tabs = this.collectLastReportData(hosts);
    reportStatus = this.lastReportStatus(hosts[activeHostId]);

    if (hosts[activeHostId]['progress_report']) {
      report = hosts[activeHostId]['progress_report'];
    }

    if ((deploymentState != 'new') && (deploymentState != 'finished')) {
      setTimeout(() => {
        loadReportData(reportDataUrl, appInstanceId);
      }, 5000);
    }

    return (
      <span>
        <div className="deploy_status">
          <div>
            <div className="deploy_status_head">Host deployment state</div>
            <div className="deploy_status_content">
              { (deploymentState != 'new' && deploymentState != 'finished') ? (<span><Spinner loading size='sm' /> &nbsp;</span>) : (<span></span>) }
              { deploymentState }
            </div>
          </div>
          <div>
            <div className="deploy_status_head">Deployment task</div>
            <div className="deploy_status_content"><a href={ deployTaskUrl } >Last deployment task</a></div>
          </div>
          <div>
            <div className="deploy_status_head">Configuration job</div>
            <div className="deploy_status_content"><a href={ configureJobUrl }>Configuration jobs</a></div>
          </div>
        </div>
        <div className="deploy_report_hosts">
          Hosts
          <VerticalTabs id="vertical_tabs">
            {tabs}
          </VerticalTabs>
        </div>
        <div className="deploy_report_status">
          {reportStatus}
        </div>
        <div className="deploy_report_tasks">
          <ReportViewer report={report} />
        </div>
      </span>
    )};
}

ApplicationInstanceReport.defaultProps = {
  error: {},
  appInstanceName: '',
  deployTaskUrl: '',
  configureJobUrl: '',
  hosts: [],
  report: [],
  activeHostId: 0,
  deploymentState: 'unknown',
}

ApplicationInstanceReport.propTypes = {
  initApplicationInstanceReport: PropTypes.func,
  appInstanceName: PropTypes.string,
  deployTaskUrl: PropTypes.string,
  configureJobUrl: PropTypes.string,
  hosts: PropTypes.array,
  deploymentState: PropTypes.string,
  report: PropTypes.array,
  setActiveHost: PropTypes.func,
  loadReportData: PropTypes.func,
  activeHostId: PropTypes.number,
};

export default ApplicationInstanceReport;
