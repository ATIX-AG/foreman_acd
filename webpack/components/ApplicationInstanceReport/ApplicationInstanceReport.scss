@import '~@theforeman/vendor/scss/variables';

.deploy_status {
  padding-bottom: 80px;
}

.deploy_status_head {
  float:left;
  width:25%;
  height:10%;
}

.deploy_status_content {
  float:left;
  width:75%;
  height:10%;
}

.deploy_report_hosts {
  float:left;
  width:25%;
  height:100%;
}

.deploy_report_tasks {
  float:left;
  width:75%;
  min-height: 300px;
  background-color: #ecf9fe;
  padding: 10px;
  border: 1px solid #3dbff3;
}

ul.vertical-tabs-pf {
  list-style-type: none;
}

.vertical-tabs-pf-tab.active {
  background-color: $dropdown-link-hover-bg;
}

.vertical-tabs-pf-tab {
  background-color: white;
}
