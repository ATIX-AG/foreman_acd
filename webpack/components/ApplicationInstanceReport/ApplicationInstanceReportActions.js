import React from 'react';
import api from 'foremanReact/API';

import {
  actionHeaderCellFormatter,
} from 'patternfly-react';

import {
  propsToSnakeCase,
  propsToCamelCase,
} from 'foremanReact/common/helpers';

import {
  APPLICATION_INSTANCE_REPORT_INIT,
  APPLICATION_INSTANCE_REPORT_SET_ACTIVE_HOST,
  APPLICATION_INSTANCE_REPORT_LOAD_REPORT_REQUEST,
  APPLICATION_INSTANCE_REPORT_LOAD_REPORT_SUCCESS,
  APPLICATION_INSTANCE_REPORT_LOAD_REPORT_FAILURE,
} from './ApplicationInstanceReportConstants';

export const initApplicationInstanceReport = (
  hosts, deploymentState,
) => dispatch => {
  const initialState = {};

  initialState.hosts = hosts;
  initialState.deploymentState = deploymentState;

  dispatch({
    type: APPLICATION_INSTANCE_REPORT_INIT,
    payload: initialState,
  });
};

export const loadReportData = (
  reportDataUrl,
  appInstanceId,
) => dispatch => {
  dispatch({ type: APPLICATION_INSTANCE_REPORT_LOAD_REPORT_REQUEST });

  const baseUrl = reportDataUrl;
  const realUrl = baseUrl.replace("__id__", appInstanceId);

  return api
    .get(realUrl, {}, {})
    .then(({ data }) =>
      dispatch({
        type: APPLICATION_INSTANCE_REPORT_LOAD_REPORT_SUCCESS,
        payload: { ...data }
      })
    )
    .catch(error => dispatch(errorHandler(APPLICATION_INSTANCE_REPORT_LOAD_REPORT_FAILURE, error)));
};

const errorHandler = (msg, err) => {
  const error = {
    errorMsg: 'Failed to fetch data from server.',
    statusText: err,
  };
  return { type: msg, payload: { error } };
};

export const setActiveHost = (id) => ({
  type: APPLICATION_INSTANCE_REPORT_SET_ACTIVE_HOST,
  payload: { activeHostId: id },
});
