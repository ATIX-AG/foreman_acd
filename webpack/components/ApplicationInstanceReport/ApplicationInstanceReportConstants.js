export const APPLICATION_INSTANCE_REPORT_INIT = 'APPLICATION_INSTANCE_REPORT_INIT';
export const APPLICATION_INSTANCE_REPORT_SET_ACTIVE_HOST = 'APPLICATION_INSTANCE_REPORT_SET_ACTIVE_HOST';
