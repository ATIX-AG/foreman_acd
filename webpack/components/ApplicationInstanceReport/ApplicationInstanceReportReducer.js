import Immutable from 'seamless-immutable';

import {
  cloneDeep,
  findIndex,
  findLastIndex,
} from 'lodash';

import {
  APPLICATION_INSTANCE_REPORT_INIT,
  APPLICATION_INSTANCE_REPORT_SET_ACTIVE_HOST,
  APPLICATION_INSTANCE_REPORT_LOAD_REPORT_REQUEST,
  APPLICATION_INSTANCE_REPORT_LOAD_REPORT_SUCCESS,
  APPLICATION_INSTANCE_REPORT_LOAD_REPORT_FAILURE,
} from './ApplicationInstanceReportConstants';

export const initialState = Immutable({
  name: false,
  error: { errorMsg: '', status: '', statusText: '' },
});

const applicationInstanceReport = (state = initialState, action) => {
  const { payload } = action;

  switch (action.type) {
    case APPLICATION_INSTANCE_REPORT_INIT: {
      return state.merge(payload);
    }
    case APPLICATION_INSTANCE_REPORT_SET_ACTIVE_HOST: {
      return state.merge({
        loading: true,
        activeHostId: payload.activeHostId,
      })
    }
    case APPLICATION_INSTANCE_REPORT_LOAD_REPORT_REQUEST: {
      // Nothing to do
      return state;
    }
    case APPLICATION_INSTANCE_REPORT_LOAD_REPORT_SUCCESS: {
      return state.merge({
        deploymentState: payload.deployment_state,
        hosts: payload.hosts,
      });
    }
    case APPLICATION_INSTANCE_REPORT_LOAD_REPORT_FAILURE: {
      console.log("Error while loading report data: "+ payload.error);
      return state.merge({ error: payload.error});
    }
    default: {
      return state;
    }
  }
};

export default applicationInstanceReport;
