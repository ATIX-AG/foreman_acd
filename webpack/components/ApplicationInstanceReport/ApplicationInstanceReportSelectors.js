const applicationInstanceReport = state => state.foremanAcd.applicationInstanceReport;

export const selectHosts = state => applicationInstanceReport(state).hosts;
export const selectActiveHostId = state => applicationInstanceReport(state).activeHostId;
export const selectDeploymentState = state => applicationInstanceReport(state).deploymentState;
