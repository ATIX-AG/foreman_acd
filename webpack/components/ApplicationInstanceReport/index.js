import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import './ApplicationInstanceReport.scss';
import ApplicationInstanceReport from './ApplicationInstanceReport';
import * as ApplicationInstanceReportActions from './ApplicationInstanceReportActions';

import {
  selectHosts,
  selectReport,
  selectActiveHostId,
} from './ApplicationInstanceReportSelectors';

const mapStateToProps = state => ({
  hosts: selectHosts(state),
  report: selectReport(state),
  activeHostId: selectActiveHostId(state),
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(ApplicationInstanceReportActions, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ApplicationInstanceReport);

