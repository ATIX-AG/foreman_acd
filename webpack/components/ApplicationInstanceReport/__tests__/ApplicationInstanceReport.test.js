import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import ApplicationInstanceReport from '../ApplicationInstanceReport';

const noop = () => {};

const hostData = [
  {
    id: 8,
    name: 'great-web-app-db-1',
    build: false,
    hostUrl: '/hosts/great-web-app-db-1.deploy3.dev.atix',
    progress_report: [],
    powerStatusUrl: '/api/v2/hosts/great-web-app-db-1.deploy3.dev.atix/power',
  },
  {
    id: 9,
    name: 'great-web-app-web-1',
    build: false,
    hostUrl: '/hosts/great-web-app-web-1.deploy3.dev.atix',
    progress_report: [],
    powerStatusUrl: '/api/v2/hosts/great-web-app-web-1.deploy3.dev.atix/power',
  },
  {
    id: 10,
    name: 'great-web-app-web-2',
    build: false,
    hostUrl: '/hosts/great-web-app-web-2.deploy3.dev.atix',
    progress_report: [],
    powerStatusUrl: '/api/v2/hosts/great-web-app-web-2.deploy3.dev.atix/power',
  },
];

const fixtures = {
  'should render application instance report': {
    data: {
      appInstanceName: 'instance name',
      deployTaskUrl: 'deploy/task/url',
      configureJobUrl: 'configure/job/url',
      hosts: hostData,
    },
    initApplicationInstanceReport: noop,
  },
};

describe('ApplicationInstanceReport', () =>
  testComponentSnapshotsWithFixtures(ApplicationInstanceReport, fixtures));
