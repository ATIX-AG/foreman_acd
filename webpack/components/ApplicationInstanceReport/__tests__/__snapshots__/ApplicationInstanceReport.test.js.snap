// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationInstanceReport should render application instance report 1`] = `
<span>
  No host
</span>
`;
